### PR TITLE
Epoll compiled on GLIBC only supports GLIBC at runtime

### LINKED ISSUE #11701: Throwing an exception in case glibc is missing instead of segfaulting the JVM
It seems Netty's native support depends on glibc (btw is this documented?).
When glibc is missing (for example using vanilla Alpine), the JVM could end up with a crash (`SIGSEGV`), see details in this issue: https://github.com/micrometer-metrics/micrometer/issues/2776

It also seems that `Epoll.isAvailable()` returns `true` even if glibc is missing.

### Expected behavior
Throwing an exception

### Actual behavior
JVM crash

### Steps to reproduce
Using Netty native with an environment that does not have glibc, e.g.: vanilla Alpine

### Minimal yet complete reproducer code (or URL to code)
This seems to be a known issue, please let me know if you really need a reproducer.

### Netty version
`4.1.68` (latest)

### JVM version (e.g. `java -version`)
Latest 11:
```bash
 ❯ docker run --rm 'azul/zulu-openjdk-alpine:11-jre' 'sh' '-c' 'java --version'
openjdk 11.0.12 2021-07-20 LTS
OpenJDK Runtime Environment Zulu11.50+19-CA (build 11.0.12+7-LTS)
OpenJDK 64-Bit Server VM Zulu11.50+19-CA (build 11.0.12+7-LTS, mixed mode)
```
Or also latest 17:
```bash
 ❯ docker run --rm 'bellsoft/liberica-openjdk-alpine-musl:17' 'sh' '-c' 'java --version'
openjdk 17 2021-09-14 LTS
OpenJDK Runtime Environment (build 17+35-LTS)
OpenJDK 64-Bit Server VM (build 17+35-LTS, mixed mode)
```
I assume this is an issue in every JRE that is currently supported.

### OS version (e.g. `uname -a`)
Both images above have the same output:
```bash
 ❯ docker run --rm 'azul/zulu-openjdk-alpine:11-jre' 'sh' '-c' 'uname -a'
Linux 1ee8d1090f14 5.10.47-linuxkit #1 SMP Sat Jul 3 21:51:47 UTC 2021 x86_64 Linux
```
I assume this is an issue in every OS where glibc is missing.

### PR DESCRIPTION
**Pull request by chrisvest**
Motivation:
It is possible to run into JVM crashes when our native binaries are compiled for one libc but then linked with another at runtime.
There is no guarantee that their ABIs will be compatible, and this can cause all sorts of trouble.
All our Linux binaries are compiled for GLIBC, so it makes sense to check that any binary compiled for GLIBC also ends up with GLIBC at runtime.
Especially since alternative libc implementations, such as musl, are getting increasingly popular in a containerised/serverless world.

Modification:
Add a check when loading the epoll native library that verifies: if the library was compiled for GLIBC, then we must also have a GLIBC loaded at runtime.
The key is to make this check without making any libc calls in our native code.
We make this check by reading `/proc/self/maps`, which contain paths for all loaded libraries in our process.
GLIBC tends to have paths with a particular pattern to their name, so we look for that.
At compile time, we rely on `__GLIBC__` to determine if the binary is compiled for GLIBC or not.
We cannot check for Musl because they intentionally expose no equivalent define to check for.
By checking for GLIBC specifically, we also end up handling all other alternative libc implementations, and not just Musl.

Result:
Epoll is now reported as not available when running a GLIBC binary on a Musl - or any other alternative libc - based system.
Previously, epoll would be reported as available, but could cause JVM crashes.

This fixes #11701

### CODE REVIEW

**Comment by chrisvest** [comment]
As the builds correctly point out, we can't rely on file names for this check. They are too volatile, and distros are free to name them however they like.

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/Native.java [hunk 1]
```diff
@@ -55,6 +55,7 @@
  * <p>Static members which call JNI methods must be defined in {@link NativeStaticallyReferencedJniMethods}.
  */
 public final class Native {
+    private static final boolean skipMuslCheck = SystemPropertyUtil.getBoolean("io.netty.native.musl.check", false);
```
Comment (normanmaurer) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/Native.java:58]: either we should rename the variable or rename the property

**Review by normanmaurer — CHANGES_REQUESTED** [review]
one comment

**Review by normanmaurer — APPROVED** [review]

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/Native.java [hunk 2]
```diff
@@ -92,6 +98,39 @@
...
+                    BufferedReader reader = new BufferedReader(new InputStreamReader(fis));
+                    boolean muslFound = false;
+                    String line;
+                    while ((line = reader.readLine()) != null) {
+                        if (line.contains("-musl-")) {
+                            muslFound = true;
+                            break;
+                        }
+                    }
+                    if (muslFound) {
```
Comment (normanmaurer) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/Native.java:116]: nit: we could just move the whole `throw` in the if block that is contained in the while loop and so get rid of the boolean at all.

**Comment by normanmaurer** [comment]
@chrisvest ship it!

**Comment by ninja-** [comment]
does this break environments that ship glibc compat library on top of Alpine, where epoll worked just fine?

**Comment by ninja-** [comment]
see for example this popular image: https://github.com/AdoptOpenJDK/openjdk-docker/blob/master/12/jre/alpine/Dockerfile.hotspot.releases.full#L34

**Comment by chrisvest** [comment]
@ninja- Is that on Dockerhub somewhere?

**Comment by chrisvest** [comment]
It's not clear what image and tag I need to pull to get that one.

**Comment by chrisvest** [comment]
On `adoptopenjdk/openjdk11:alpine` the glibc-compat shows up as `/usr/glibc-compat/lib/libc.so.6`, so that's not a problem.

**Comment by ninja-** [comment]
Just checked my prod environment. They are both in memory, so looking at your code logic it would break this environment and disable epoll.

```
/ # cat /proc/82/maps|egrep "musl|libc.so"
7f97b70a7000-7f97b70bc000 r--p 00000000 00:350 10238988                  /lib/ld-musl-x86_64.so.1
7f97b70bc000-7f97b7104000 r-xp 00015000 00:350 10238988                  /lib/ld-musl-x86_64.so.1
7f97b7104000-7f97b713a000 r--p 0005d000 00:350 10238988                  /lib/ld-musl-x86_64.so.1
7f97b713a000-7f97b713b000 r--p 00092000 00:350 10238988                  /lib/ld-musl-x86_64.so.1
7f97b713b000-7f97b713c000 rw-p 00093000 00:350 10238988                  /lib/ld-musl-x86_64.so.1
7f97e9a91000-7f97e9ab6000 r--p 00000000 00:350 10239397                  /usr/glibc-compat/lib/libc.so.6
7f97e9ab6000-7f97e9c01000 r-xp 00025000 00:350 10239397                  /usr/glibc-compat/lib/libc.so.6
7f97e9c01000-7f97e9c4b000 r--p 00170000 00:350 10239397                  /usr/glibc-compat/lib/libc.so.6
7f97e9c4b000-7f97e9c4c000 ---p 001ba000 00:350 10239397                  /usr/glibc-compat/lib/libc.so.6
7f97e9c4c000-7f97e9c4f000 r--p 001ba000 00:350 10239397                  /usr/glibc-compat/lib/libc.so.6
7f97e9c4f000-7f97e9c52000 rw-p 001bd000 00:350 10239397                  /usr/glibc-compat/lib/libc.so.6
```

**Comment by ninja-** [comment]
Thanks for reverting.
Honestly, I haven't even seen a Java distribution that works on musl, is that even a thing these days?
Every image I've seen had a glibc compatibility layer.

**Comment by chrisvest** [comment]
I think Java 16 or 17 added support in upstream openjdk. Haven't looked into the details, but there are builds of it available.

**Pull request merged**